android-pdf-viewer/src/main/java/com/github/barteksc/pdfviewer/annotation/core/pdf/PdfUtil.kt

```diff
@@ -0,0 +1,327 @@
+            return annotationsList
+        } catch (e: Exception) {
+            e.message?.let { logError(TAG, it) }
+            e.printStackTrace()
+            return emptyList()
+        }
+    }
+
+    private fun getExtractedSquareAnnotation(
+        annotation: PdfDictionary,
+        llx: Float,
+        lly: Float,
+        urx: Float,
+        ury: Float
+    ): Annotation {
+        val xBottomLeftPoint = llx
+        val yBottomLeftPoint = lly
+        val bottomLeftPoint = PointF(xBottomLeftPoint, yBottomLeftPoint)
+        val xTopRightPoint = urx
+        val yTopRightPoint = ury
+        val topRightPoint = PointF(xTopRightPoint, yTopRightPoint)
+
+        // from the extracted coordinates, calculate the rest
+        val squareAnnotationPoints =
+            generateRectangleCoordinates(bottomLeftPoint, topRightPoint)
+
+        val relationsArray: PdfArray? =
+            annotation.getAsArray(PdfName("relations"))
+
+        return Annotation(
+            AnnotationType.SQUARE.name,
+            squareAnnotationPoints,
+            relations = getExtractedRelations(relationsArray)
+        )
+    }
+
+    private fun getExtractedRelations(relationsArray: PdfArray?): Relations? {
+        val documentations = mutableListOf<Documentation>()
+        if (relationsArray != null) {
+            for (j in 0 until relationsArray.size()) {
+                val documentationDict: PdfDictionary =
+                    relationsArray.getAsDict(j)
+                val schemaId: PdfNumber? =
+                    documentationDict.getAsNumber(PdfName("schemaId"))
+                val documentId: PdfString? =
+                    documentationDict.getAsString(PdfName("documentId"))
+
+                if (schemaId != null && documentId != null) {
+                    documentations.add(
+                        Documentation(
+                            schemaId.intValue().toLong(),
+                            documentId.toString()
+                        )
+                    )
+                }
+            }
+        } else {
+            return null
+        }
+        return Relations(documentations)
+    }
+
+    /** Map annotations to shapes,
+     *  using the page height when converting between PDF space and image space */
+    private fun getShapesFor(
+        pdfAnnotations: List<Annotation>,
+        pageHeight: Int
+    ): List<Shape> {
+        // convert annotation to shape
+        val shapes = pdfAnnotations.map { annotation ->
+            when (annotation.type) {
+                AnnotationType.SQUARE.name -> return@map annotation.toRectangleShape(pageHeight)
+                else -> throw Exception("Annotation is not recognised")
+            }
+        }
+        return shapes
+    }
+
+    /** Use the passed PDF file path to map the PDF page to PNG image and
+     *  map PDF annotations to image shapes, save the PNG image to the given output directory
+     */
+    @JvmStatic
+    @Throws(IOException::class)
```

Comment:
This states that it throws an exception but that exception is not handled in the usages. Handle this here and return a nullable result instead to make this simpler for the callers

Reply:
I had tried to not throw exception and return nullable but then when returning from measureLib back to the PDF activity, the annotations weren't drawn to the pdf. I handled the exceptions in the usages in BB

Reply:
>  but then when returning from measureLib back to the PDF activity, the annotations weren't drawn to the pdf

DId you debug why? Catching any exception here and returning null and handling the exception on bb side is the same so what are we missing here?

Handling the exception here and returning null has the advantage of not delegating the exception handling to the caller.

Reply:
Will try with nullable again